README.md

```diff
@@ -29,122 +37,1058 @@ The library design is inspired by the work in [Exploring HTN Planners through ex
+	// This task will be chosen only if one of the keys is smaller than the target
+	condition: (state, { target }) =>
+		Object.keys(state.counters).some(
+			(k) => state.counters[k] < target.counters[k],
+		),
+	effect: (state, { target }) => {
+		// We find the first counter below the target, we know it exists because of
+		// the condition so we use the non-null assertion (!) at the end
+		const key = Object.keys(state._.counters).find(
+			(k) => state._.counters[k] < target.counters[k],
+		)!;
+
+		// Update the changed counter
+		state._.counters[key]++;
+	},
+	description: '+1',
+});
+```
+
+The above code achieves the same as the single counter `plusOne` (increase one of the counters), but is a bit convoluted, and we cannot tell from the description which counter will be increased. We can improve this setting the `lens` property when defining the action.
+
+```typescript
+const plusOne = Task.of<System>().from({
+	// The lens defines the part of the state where this task is applicable
+	// this property defaults to '/', meaning the root object
+	lens: '/counters/:counterId',
+	condition: (counter, { target }) => counter < target,
+	effect: (counter) => ++counter._,
+	description: ({ counterId }) => `${counterId} + 1`,
+});
+```
+
+We are again introducing a few new things here, but hopefully the code above is relatively intuitive. This declares `plusOne` as a task on the `System` type, but that acts on a specific counter, specified by the lens property. Since we don't know a priori which counter this will apply to, we replace the name of the counter with the placeholder variable, `:counterId`. This variable is passed as part of the context to the task constructor properties, we can see this in the use of the value by the description which now we see that it can also be a function acting on the context.
+
+Note also that instead of the usual `Task.from` we preface the `from` call with a `Task.of` with a generic type argument. This is because of a limitation in typescript that does not allow it to [perform partial inference of generic types](https://github.com/microsoft/TypeScript/issues/10571). We'll see later how we can shorten this call.
+
+Now we can go a bit deeper on the context object, we mentioned before that it contains the context from the task selection process during planning. More specifically, the context object is composed by
+
+- A `target` property, with the expected target value for the part of the state that the task acts on. This will have the same type as the first argument of the function, which will match the type pointed by the `lens` property. We'll see when we talk about [operations](#operations) that this property is not present for delete or wildcard (`*`) operation.
+- A `system` property, providing a copy of the global state at the moment the task is used. This can be used, for instance, to define tasks with conditions on parts of the system that are unrelated to the sub-object pointed to by the lens.
```

Comment:
This is easily abusable, right? I can imagine agents getting hard to understand and maintain if this is used too much. (Not unlike global variables, which are the obvious comparison.) So:

* Is this here because of some perceived use case, or "just in case"?
* Should we at least warn users against it?

Reply:
I guess it depends on the state model, but I know of a few use cases. For instance, for service orchestration, services can have a `dependsOn` property. I don't know if there is a good way to define conditions for that property without the service task being able to access information about the rest of the services. The example for WifiConnect at the bottom of the readme has other instance of that.

I was hesitant to add the system property for the same reasons you point out, but I decided it was necessary.

We can add some note regarding the use of the `system` property, something on the lines "if you find that you are using the system property too much, consider modifying the model"

Reply:
I added a note, let me know what you think